packages/governance/src/governor/extensions/governor_timelock_execution.cairo

```diff
@@ -0,0 +1,295 @@
+    use crate::governor::extensions::interface::ITimelocked;
+    use crate::governor::interface::ProposalState;
+    use crate::timelock::interface::{ITimelockDispatcher, ITimelockDispatcherTrait};
+    use openzeppelin_introspection::src5::SRC5Component;
+    use starknet::ContractAddress;
+    use starknet::account::Call;
+    use starknet::storage::{Map, StorageMapReadAccess, StorageMapWriteAccess};
+    use starknet::storage::{StoragePointerReadAccess, StoragePointerWriteAccess};
+
+    type ProposalId = felt252;
+    type TimelockProposalId = felt252;
+
+    #[storage]
+    pub struct Storage {
+        pub Governor_timelock_controller: ContractAddress,
+        pub Governor_timelock_ids: Map<ProposalId, TimelockProposalId>
+    }
+
+    #[event]
+    #[derive(Drop, starknet::Event)]
+    pub enum Event {
+        TimelockUpdated: TimelockUpdated,
+    }
+
+    /// Emitted when the timelock controller used for proposal execution is modified.
+    #[derive(Drop, starknet::Event)]
+    pub struct TimelockUpdated {
+        pub old_timelock: ContractAddress,
+        pub new_timelock: ContractAddress
+    }
+
+    pub mod Errors {
+        pub const INVALID_TIMELOCK_CONTROLLER: felt252 = 'Invalid timelock controller';
+    }
+
+    //
+    // Extensions
+    //
+
+    pub impl GovernorExecution<
+        TContractState,
+        +GovernorComponent::HasComponent<TContractState>,
+        +GovernorComponent::GovernorSettingsTrait<TContractState>,
+        +GovernorComponent::GovernorCountingTrait<TContractState>,
+        +GovernorComponent::GovernorVotesTrait<TContractState>,
+        +SRC5Component::HasComponent<TContractState>,
+        impl GovernorCoreExecution: HasComponent<TContractState>,
+        +Drop<TContractState>
+    > of GovernorComponent::GovernorExecutionTrait<TContractState> {
+        /// See `GovernorComponent::GovernorExecutionTrait::state`.
+        fn state(
+            self: @GovernorComponentState<TContractState>, proposal_id: felt252
+        ) -> ProposalState {
+            let current_state = self._state(proposal_id);
+
+            if current_state != ProposalState::Queued {
+                return current_state;
+            }
+
+            let contract = self.get_contract();
+            let this_component = GovernorCoreExecution::get_component(contract);
+
+            let queue_id = this_component.Governor_timelock_ids.read(proposal_id);
+            let timelock_dispatcher = this_component.get_timelock_dispatcher();
+
+            if timelock_dispatcher.is_operation_pending(queue_id) {
+                ProposalState::Queued
+            } else if timelock_dispatcher.is_operation_done(queue_id) {
+                // This can happen if the proposal is executed directly on the timelock.
+                ProposalState::Executed
+            } else {
+                // This can happen if the proposal is canceled directly on the timelock.
+                ProposalState::Canceled
+            }
```

Comment:
It should be easier to read like this (pending checks to states and we don't have to compare with OperationState members), and since get_operation_state is internal to timelock, the update will only save us one internal function JUMP, not a external call, then is not worth losing clarity on this case.

Reply:
Correct me if I'm wrong, but in the case of `Executed` or `Canceled` proposal state we'll have to make 2 external calls to timelock contract (`is_operation_pending` + `is_operation_done`) instead of a single `get_operation_state` call

Reply:
You are right, I was thinking of two calls to get_operation_state, but there's no need for that. Updated!